### ISSUE TITLE
[bug] Bug in MADE implementation used for IAF

### ISSUE THREAD

**Issue by stefanwebb** [uber/pyro#1256]
I think I've found a bug in the MADE code in [this part of the code](https://github.com/uber/pyro/blob/dev/pyro/nn/auto_reg_nn.py#L72).

The expression:
`torch.multinomial(torch.tensor(input_dim - 1).fill_(1.).float(), num_samples=hidden_dim, replacement=True)` 
always returns a vector of zeros for me (not sure why, perhaps because of in-place operations?), whereas it works as it ought if I change it to,
`torch.multinomial(torch.ones(input_dim - 1), num_samples=hidden_dim, replacement=True)`
outputting a random vector of random integers on `0:(input_dim-1)`.

I'm using the nightly build of Pytorch from 7/5/2018, and Python 3.6.5 on Ubuntu

**Comment by stefanwebb** [comment]
I'm reworking the Pyro IAF code for my own experiments and adding a lot of features - would this be welcomed if I submitted it as a pull request?

I'm adding:

- using multi-layer MADEs with skip connections
- conditional MADEs for conditional densities
- explicitly calculating the IAF inverse
- multiple steps of normalizing flow
- MAF and NAF

**Comment by karalets** [comment]
That would be great!

**Comment by martinjankowiak** [comment]
@stefanwebb 

that would be awesome. in doing so you might consider doing the following:

1) open up an issue to highlight/discuss some of the design choices you're making. it's possible
that some group discussion and targeted comments could improve upon your initial design

2) when you're ready to submit code, submit a series of PRs (each adding one or two discrete
new features) instead of one giant PR. note that doing 1) before 2) may help make the code review
process smoother

**Comment by fritzo** [comment]
Oh right, that should have been `torch.Tensor(input_dim - 1)` rather than `torch.tensor(input_dim - 1)`. The former creates an empty tensor of shape `(input_dim-1,)` whereas the latter creates a scalar `LongTensor` of shape `()` filled with the value `input_dim - 1`.

**Comment by stefanwebb** [comment]
Ah of course!

**Comment by neerajprad** [comment]
@stefanwebb - Now that the master task is being tracked in #1260, can we close this issue, or does the bug still exist?

**Comment by stefanwebb** [comment]
@neerajprad yes, I'll close this issue now! That bug was fixed in the rewrite of the MADE class
